nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+      attachedAssets: {
+        NEO: '100',
+        GAS: '0.0001',
+      },
+      triggerContractVerification: true,
+    },
+    {
+      scriptHash: '505663a29d83663a838eee091249abd167e928f5',
+      operation: 'purchaseTicket',
+      arguments: [
+        {
+          type: 'number',
+          value: '10'
+        }
+      ],
+    }
+  ],
+  fee: '0.001',
+  network: 'TestNet',
+  broadcastOverride: false,
+  txHashAttributes: [
+    {
+      type: 'Boolean',
+      value: true,
+      txAttrUsage: 'Hash1'
+    }
+  ]
+})
+.then(({txid, nodeUrl}: InvokeOutput) => {
+  console.log('Invoke transaction success!');
+  console.log('Transaction ID: ' + txid);
+  console.log('RPC node URL: ' + nodeUrl);
+})
+.catch(({type: string, description: string, data: any}) => {
+  switch(type) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case RPC_ERROR:
+      console.log('There was an error when broadcasting this transaction to the network.');
+      break;
+    case CANCELED:
+      console.log('The user has canceled this transaction.');
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request
+* Validate and format the inputs
+** Script hashes is valid
+** Format parameters
+*** If `byteArray`, convert to appropriate hexstring based on assumed format
+* Batch invoke inputs into script, and handle assets to each invoke as outputs
+* Set script transaction attribute 0x20 according to the following conditions
+** If triggerContractVerification is set to true, set 0x20 to scriptHash of the contract being invoked
+** If there is no fee, attachedAssets, or 'assetIntentOverrides', set 0x20 to the users address
+** If there are assetIntentOverrides but none of the inputs belong to the user address, set 0x20 to user address
+* If the "broadcastOverride" input argument is set to True, return the serialized signed transaction to the dApp
+* Else broadcast the RPC request
+* Return the transaction Id and the url of the rpc node which the tx was broadcast to
+
+Example
+<pre>
+{
+  txid: 'ed54fb38dff371be6e3f96e4880405758c07fe6dd1295eb136fe15f311e9ff77',
+  nodeUrl: 'http://seed7.ngd.network:10332',
+}:
+</pre>
+
+
+====signMessage====
+Signs a provided messaged with an account selected by user. A randomized salt prefix is added to the input string before it is signed, and it is encased in a non-executable transaction before signed. This ensures allow compatibility with Ledger devices.
+
+
+'''Method Interface'''
+
+<pre>
+function signMessage({message: string}): Promise<SignedMessage>
+</pre>
+
+
+'''Input arguments'''
+
+<pre>
+interface SignMessageArgs {
+  message: string; // Arbitrary message to sign
+}
+</pre>
+
+
+'''Success return value'''
+
+<pre>
+interface SignedMessage {
+  publicKey: string; // Public key of account that signed message
+  message: string; // Original message signed
+  salt: string; // Salt added to original message as prefix, before signing
```

Comment:
I don't get this `salt` here... is it added to the script itself, like nep5 transfers, to avoid hash collisions (and signature exposure)? Or is it related to the signing process itself on secp256r1... if it's for signing itself, perhaps this decision must not be made here by the user, because if it repeats signature "by mistake", and publishes it, his signature will be forever lost. It's very dangerous indeed! One small mistake, all funds lost.
Usually, what crypto libraries do is: (A) generating cryptographically safe numbers, by connecting to local host system random() generator (like C#), or (B) generating a deterministic number based on private key and message itself, so that no bytes are exposed from private key (some still don't think it's safe, but looks like it is...), which is what neon-js does (it uses deterministic signing from another js crypto library).
Anyway, it could be dangerous to allows users  to mess up with this here.

Reply:
This is not set by the user, but by the wallet provider. The purpose of signMessage is not to sign arbitrary transactions, but rather to sign arbitrary messages. It's original intention was to also prevent any dapp from attempting to use this method to have a user sign an arbitrary message and also to prevent the signed output of the same input from being identical. Since the original proposal, this method has enclosed the message in an empty transaction to allow for signing by ledger devices, and the benefit now is limited to the second benefit only.

Reply:
I still dont get it @nickfujita but if you say its safe, I believe you. Probably I'm not seeing the whole picture.